src/Microsoft.TestPlatform.CrossPlatEngine/TestSession/ProxyTestSessionManager.cs

```diff
@@ -382,6 +384,35 @@ private void DisposeProxies()
             _proxyMap.Clear();
         }
     }
+
+    private bool CheckRunSettingsAreCompatible(string requestRunSettings)
```

Comment:
perhaps this should be converted to a void and called ProcessRunSettings? and throw if the runsettings are invalid?
might be slightly cleaner than having a bool that underneath mutates the classes fields

Reply:
Oh, but we don't actually mutate the class fields. The only thing that can make it look like we are is this piece of code, but this is only for caching purposes so that we don't have to reparse the test session run settings (which never change) each time we have to check if the new runsettings match the old ones:
```
if (_runSettingsEnvironmentVariables.Count == 0)
{
    _runSettingsEnvironmentVariables = InferRunSettingsHelper.GetEnvironmentVariables(
            _testSessionCriteria.RunSettings)
        ?? _runSettingsEnvironmentVariables;
}
```

Also, I don't feel like `ProcessRunSettings` is an accurate name. All this function does is to literally check if the test session run settings match the new run settings so a testhost from the current session may be reused, hence coming up with the current naming.

Reply:
but the cached value is what's being mutated. thus calling CheckRunSettingsAreCompatible() twice with the same runsettings would return false and then return true. usually for bool returning methods I'd expect that this is a query and so it should return the same result if called twice

Reply:
That's not the case, `CheckRunSettingsAreCompatible()` will always return the same value given the same input. I wrote a test in order to confirm and also have this as a check against future changes.

That being said, I did some refactoring and hid the caching part behind a property to make the `CheckRunSettingsAreCompatible()` method easier to follow.